<center> <font size=6> Artificial neural network: Is that picture an advertisement or not? </font></center>

Load the data.

In [1]:
import pandas as pd
df=pd.read_csv("advertisement.csv",header=None, low_memory=False)
df

0           1         2     3     4     5     6     7     8     \
0     125.000000  125.000000  1.000000     1     0     0     0     0     0   
1      57.000000  468.000000  8.210500     1     0     0     0     0     0   
2      33.000000  230.000000  6.969600     1     0     0     0     0     0   
3      60.000000  468.000000  7.800000     1     0     0     0     0     0   
4      60.000000  468.000000  7.800000     1     0     0     0     0     0   
...          ...         ...       ...   ...   ...   ...   ...   ...   ...   
3274  170.000000   94.000000  0.552900     0     0     0     0     0     0   
3275  101.000000  140.000000  1.386100     1     0     0     0     0     0   
3276   23.000000  120.000000  5.217300     1     0     0     0     0     0   
3277   64.021886  155.344828  3.911953     1     0     0     0     0     0   
3278   40.000000   40.000000  1.000000     1     0     0     0     0     0   

      9     ...  1549  1550  1551  1552  1553  1554  1555  1556  1557    1558  
0        0  ...     0     0     0     0     0     0     0     0     0     ad.  
1        0  ...     0     0     0     0     0     0     0     0     0     ad.  
2        0  ...     0     0     0     0     0     0     0     0     0     ad.  
3        0  ...     0     0     0     0     0     0     0     0     0     ad.  
4        0  ...     0     0     0     0     0     0     0     0     0     ad.  
...    ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   ...     ...  
3274     0  ...     0     0     0     0     0     0     0     0     0  nonad.  
3275     0  ...     0     0     0     0     0     0     0     0     0  nonad.  
3276     0  ...     0     0     0     0     0     0     0     0     0  nonad.  
3277     0  ...     0     0     0     0     0     0     0     0     0  nonad.  
3278     0  ...     0     0     0     0     0     0     0     0     0  nonad.  

[3279 rows x 1559 columns]

The data has been cleaned on Excel. There are many missing values in columns 0, 1,,2,3. I replaced the missing values on columns 0, 1,2 with the column mean and replaced the missing values in column 3 with the mode. Python's "mean" and "mode" methods could not handle missing values. However, Excel's "AVERAGE" and "MODE" functions automatically ignore any non-numeric values and calculate the mean and mode, respectively, on numeric values. 

It is also a lot quicker to replace the question marks on Excel also. I used nested the "ISNUMBER" function inside the logical "IF" function and just clicked to fill the entire columns. 

Here is another reason to clean the data on Excel. "df.iloc[:,0]==" ?" returned FALSE for all rows, with or without the space in front of the question mark. All in all, for this dataset, it is a lot quicker to clean it on Excel and upload the cleaned dataset to the Python environment.

Separate the potential predictors from the target.

In [2]:
X=df.drop(columns=1558,axis=1)
X

0           1         2     3     4     5     6     7     8     \
0     125.000000  125.000000  1.000000     1     0     0     0     0     0   
1      57.000000  468.000000  8.210500     1     0     0     0     0     0   
2      33.000000  230.000000  6.969600     1     0     0     0     0     0   
3      60.000000  468.000000  7.800000     1     0     0     0     0     0   
4      60.000000  468.000000  7.800000     1     0     0     0     0     0   
...          ...         ...       ...   ...   ...   ...   ...   ...   ...   
3274  170.000000   94.000000  0.552900     0     0     0     0     0     0   
3275  101.000000  140.000000  1.386100     1     0     0     0     0     0   
3276   23.000000  120.000000  5.217300     1     0     0     0     0     0   
3277   64.021886  155.344828  3.911953     1     0     0     0     0     0   
3278   40.000000   40.000000  1.000000     1     0     0     0     0     0   

      9     ...  1548  1549  1550  1551  1552  1553  1554  1555  1556  1557  
0        0  ...     0     0     0     0     0     0     0     0     0     0  
1        0  ...     0     0     0     0     0     0     0     0     0     0  
2        0  ...     0     0     0     0     0     0     0     0     0     0  
3        0  ...     0     0     0     0     0     0     0     0     0     0  
4        0  ...     0     0     0     0     0     0     0     0     0     0  
...    ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
3274     0  ...     0     0     0     0     0     0     0     0     0     0  
3275     0  ...     0     0     0     0     0     0     0     0     0     0  
3276     0  ...     0     0     0     0     0     0     0     0     0     0  
3277     0  ...     0     0     0     0     0     0     0     0     0     0  
3278     0  ...     0     0     0     0     0     0     0     0     0     0  

[3279 rows x 1558 columns]

In [3]:
y=pd.DataFrame(df.iloc[:,1558])
y

1558
0        ad.
1        ad.
2        ad.
3        ad.
4        ad.
...      ...
3274  nonad.
3275  nonad.
3276  nonad.
3277  nonad.
3278  nonad.

[3279 rows x 1 columns]

Feature selection.

In [4]:
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.feature_selection import SelectFromModel

clf = ExtraTreesClassifier(n_estimators=50)
clf = clf.fit(X, y)
scores=pd.DataFrame(clf.feature_importances_).rename(columns={0:"scores"})
scores=scores.sort_values(by=["scores"],ascending=False).nlargest(50,"scores")
scores

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:6: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  


scores
1     0.055446
1399  0.045164
1243  0.041908
1229  0.030795
351   0.029218
1483  0.025818
968   0.019719
1435  0.018061
2     0.017005
0     0.016680
1047  0.015222
1394  0.011606
874   0.010691
1108  0.010674
1153  0.010298
398   0.009928
1278  0.009926
1276  0.009714
1455  0.009335
1247  0.008974
1362  0.008350
1116  0.008169
1344  0.007791
1385  0.007493
1424  0.007375
983   0.007250
1143  0.007140
860   0.007028
1022  0.006561
1040  0.006465
1445  0.006281
307   0.005919
3     0.005878
317   0.005832
1309  0.005820
1218  0.005788
167   0.005686
1118  0.005683
187   0.005558
508   0.005383
1271  0.005242
1373  0.005052
1207  0.005009
1293  0.004971
1198  0.004760
1254  0.004544
1326  0.004440
175   0.004324
1179  0.004285
1359  0.004283

Extract the column indices of the selected features.

In [5]:
import numpy as np
np.asarray(scores.index)

array([   1, 1399, 1243, 1229,  351, 1483,  968, 1435,    2,    0, 1047,
       1394,  874, 1108, 1153,  398, 1278, 1276, 1455, 1247, 1362, 1116,
       1344, 1385, 1424,  983, 1143,  860, 1022, 1040, 1445,  307,    3,
        317, 1309, 1218,  167, 1118,  187,  508, 1271, 1373, 1207, 1293,
       1198, 1254, 1326,  175, 1179, 1359])

Form the predictor set.

In [6]:
X=X.iloc[:,[1, 1399, 1243, 1229,  351, 1483,  968, 1435,    2,    0, 1047,
       1394,  874, 1108, 1153,  398, 1278, 1276, 1455, 1247, 1362, 1116,
       1344, 1385, 1424,  983, 1143,  860, 1022, 1040, 1445,  307,    3,
        317, 1309, 1218,  167, 1118,  187,  508, 1271, 1373, 1207, 1293,
       1198, 1254, 1326,  175, 1179, 1359]]
X

1     1399  1243  1229  351   1483  968   1435      2     \
0     125.000000     0     1     0     0     1     0     1  1.000000   
1     468.000000     0     1     0     0     0     0     0  8.210500   
2     230.000000     0     0     1     0     0     0     0  6.969600   
3     468.000000     0     0     1     0     1     0     0  7.800000   
4     468.000000     0     0     1     0     0     0     0  7.800000   
...          ...   ...   ...   ...   ...   ...   ...   ...       ...   
3274   94.000000     0     0     0     0     0     0     0  0.552900   
3275  140.000000     0     0     0     0     0     0     0  1.386100   
3276  120.000000     0     0     0     0     0     0     0  5.217300   
3277  155.344828     0     0     0     0     0     0     0  3.911953   
3278   40.000000     0     0     0     0     0     0     0  1.000000   

            0     ...  1271  1373  1207  1293  1198  1254  1326  175   1179  \
0     125.000000  ...     0     0     0     0     0     0     0     0     0   
1      57.000000  ...     0     0     0     0     0     0     0     0     0   
2      33.000000  ...     0     0     0     0     0     0     0     0     0   
3      60.000000  ...     0     0     0     0     0     0     0     0     0   
4      60.000000  ...     0     0     0     0     0     0     0     0     0   
...          ...  ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
3274  170.000000  ...     0     0     0     0     0     0     0     0     0   
3275  101.000000  ...     0     0     0     0     0     0     0     0     0   
3276   23.000000  ...     0     0     0     0     0     0     0     0     0   
3277   64.021886  ...     0     0     0     0     0     0     0     0     0   
3278   40.000000  ...     0     0     0     0     0     0     0     0     0   

      1359  
0        0  
1        0  
2        0  
3        1  
4        1  
...    ...  
3274     0  
3275     0  
3276     0  
3277     0  
3278     0  

[3279 rows x 50 columns]

One-hot encode the target variable.

In [7]:
from sklearn.preprocessing import OneHotEncoder
enc = OneHotEncoder(handle_unknown='ignore')
y=pd.DataFrame(enc.fit_transform(y).toarray())
y.head()

0    1
0  1.0  0.0
1  1.0  0.0
2  1.0  0.0
3  1.0  0.0
4  1.0  0.0

Split the data for training and testing.

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)
print(X_train.shape)
print(X_test.shape)

(2459, 50)
(820, 50)


Build the neural network model.

In [13]:
length=X_train.shape[1]
num_classes=y_test.shape[1]

import keras
from keras.models import Sequential
from keras.layers import Dense

# define classification model

def classification_model():
    # create model
    model = Sequential()
    model.add(Dense(length, activation='relu', input_shape=(length,)))
    model.add(Dense(65,activation='relu'))
    model.add(Dense(65,activation="relu"))
    model.add(Dense(65,activation="relu"))
    model.add(Dense(65,activation="relu"))
    model.add(Dense(65,activation="tanh"))
    model.add(Dense(num_classes, activation='softmax'))
    
    
    # compile model
    model.compile(optimizer='Adam', loss='categorical_crossentropy', metrics=['accuracy'])
    return model

Train the model on train data and test it on test data.

In [14]:
model=classification_model()
model.fit(X_train,y_train,validation_data=(X_test,y_test),epochs=20, verbose=2)
model.evaluate(X_test,y_test,verbose=0)

Train on 2459 samples, validate on 820 samples
Epoch 1/20
 - 3s - loss: 0.3016 - acc: 0.8987 - val_loss: 0.3600 - val_acc: 0.9049
Epoch 2/20
 - 2s - loss: 0.2708 - acc: 0.9211 - val_loss: 0.3038 - val_acc: 0.8927
Epoch 3/20
 - 1s - loss: 0.2714 - acc: 0.9142 - val_loss: 0.2968 - val_acc: 0.8963
Epoch 4/20
 - 2s - loss: 0.2588 - acc: 0.9118 - val_loss: 0.3254 - val_acc: 0.9037
Epoch 5/20
 - 1s - loss: 0.2337 - acc: 0.9244 - val_loss: 0.2406 - val_acc: 0.9073
Epoch 6/20
 - 2s - loss: 0.1938 - acc: 0.9394 - val_loss: 0.2222 - val_acc: 0.9146
Epoch 7/20
 - 2s - loss: 0.1905 - acc: 0.9418 - val_loss: 0.1837 - val_acc: 0.9317
Epoch 8/20
 - 2s - loss: 0.1477 - acc: 0.9524 - val_loss: 0.1678 - val_acc: 0.9378
Epoch 9/20
 - 1s - loss: 0.1417 - acc: 0.9573 - val_loss: 0.1459 - val_acc: 0.9512
Epoch 10/20
 - 1s - loss: 0.1444 - acc: 0.9618 - val_loss: 0.1476 - val_acc: 0.9524
Epoch 11/20
 - 1s - loss: 0.1248 - acc: 0.9650 - val_loss: 0.1395 - val_acc: 0.9573
Epoch 12/20
 - 1s - loss: 0.1247 - acc

[0.12454447037563092, 0.9658536585365853]

It is interesting that I experimented using 50, 100 or 200 predictor features and there was no significant difference on final accuracy on test data. In all cases, the final accuracy on test data ranged rom 95.5% to 96.5%.

In [15]:
from sklearn.metrics import roc_auc_score
print("AUC: ",roc_auc_score(y_test,model.predict(X_test)))

AUC:  0.9759233286582516


Pretty good.